### PR TITLE
[FLINK-1605] Bundle all hadoop dependencies and shade guava away

### PR DESCRIPTION
**Pull request by rmetzger**
I don't have a working eclipse setup right now, so I didn't test this change with eclipse.

I would be very interested in some feedback from there.
Its working with IntelliJ.

### CODE REVIEW

**Comment by StephanEwen** [comment]
Looks good, but also like a lot of magic. Can you briefly summarize here (and copy it to the wiki) how the setup works? I am especially confused by the extra YARN shaded versions (two of them).


Some remaining questions:
  - Why give up the `ship` directory? I think it was good to differentiate between `ship` and `lib`.
  - What is included in the final flink-dist fat jar? All streaming connectors? The hbase connector and hadoop compatibility?

**Comment by rmetzger** [comment]
I know, the stuff is very complicated.

I've documented the setup here: https://cwiki.apache.org/confluence/display/FLINK/Dependency+Shading+in+Flink
Please let me know if there are details missing in the text.

Regarding `ship` and `lib`: If you want I can undo the change (I've spend some time implementing the feature, so I'm actually in favor of keeping it).
I've removed it because I thought non-yarn users might be confused by this directory (i combined the hadoop2 and hadoop2-yarn binaries).

The final flink-dist fat jar contains:
 * the streaming connectors (my java6 pr will remove them from there
 * no hbase
 * no hadoopcompat

please don't merge the pull request! I've found an issue while writing the documentation. I'm currently verifying my fix.

**Comment by rmetzger** [comment]
The pull request has been rebased to master.

I'm looking for some feedback for this change.

File: flink-shaded-hadoop/flink-shaded-hadoop1/pom.xml [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+	<packaging>jar</packaging>
+
+	<dependencies>
+		<dependency>
+			<groupId>org.apache.hadoop</groupId>
+			<artifactId>hadoop-core</artifactId>
+			<version>${hadoop.version}</version>
+			<exclusions>
+				<exclusion>
+					<groupId>asm</groupId>
```
Comment (StephanEwen) [inline flink-shaded-hadoop/flink-shaded-hadoop1/pom.xml]: I this groupId correct, or should it be `org.ow2.asm` ?

File: flink-shaded-hadoop/flink-shaded-hadoop2/pom.xml [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+	<packaging>jar</packaging>
+
+	<dependencies>
+		<dependency>
+			<groupId>org.apache.hadoop</groupId>
+			<artifactId>hadoop-common</artifactId>
+			<version>${hadoop.version}</version>
+			<exclusions>
+				<exclusion>
+					<groupId>asm</groupId>
```
Comment (StephanEwen) [inline flink-shaded-hadoop/flink-shaded-hadoop2/pom.xml]: I this groupId correct, or should it be `org.ow2.asm` ?

File: flink-shaded-hadoop/flink-shaded-hadoop2/pom.xml [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+			</exclusions>
+		</dependency>
+
+		<dependency>
+			<groupId>org.apache.hadoop</groupId>
+			<artifactId>hadoop-hdfs</artifactId>
+			<version>${hadoop.version}</version>
+			<exclusions>
+				<exclusion>
+					<groupId>asm</groupId>
```
Comment (StephanEwen) [inline flink-shaded-hadoop/flink-shaded-hadoop2/pom.xml]: I this groupId correct, or should it be `org.ow2.asm` ?

File: flink-shaded-hadoop/flink-shaded-hadoop2/pom.xml [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+			</exclusions>
+		</dependency>
+
+		<dependency>
+			<groupId>org.apache.hadoop</groupId>
+			<artifactId>hadoop-mapreduce-client-core</artifactId>
+			<version>${hadoop.version}</version>
+			<exclusions>
+				<exclusion>
+					<groupId>asm</groupId>
```
Comment (StephanEwen) [inline flink-shaded-hadoop/flink-shaded-hadoop2/pom.xml]: I this groupId correct, or should it be `org.ow2.asm` ?

File: flink-shaded-hadoop/flink-shaded-include-yarn/pom.xml [hunk 5]
```diff
@@ -0,0 +1,286 @@
...
+	<packaging>jar</packaging>
+
+	<dependencies>
+		<dependency>
+			<groupId>org.apache.hadoop</groupId>
+			<artifactId>hadoop-common</artifactId>
+			<version>${hadoop.version}</version>
+			<exclusions>
+				<exclusion>
+					<groupId>asm</groupId>
```
Comment (StephanEwen) [inline flink-shaded-hadoop/flink-shaded-include-yarn/pom.xml]: I this groupId correct, or should it be `org.ow2.asm` ?

File: flink-shaded-hadoop/flink-shaded-include-yarn/pom.xml [hunk 6]
```diff
@@ -0,0 +1,286 @@
...
+			</exclusions> 
+		</dependency>
+
+		<dependency>
+			<groupId>org.apache.hadoop</groupId>
+			<artifactId>hadoop-hdfs</artifactId>
+			<version>${hadoop.version}</version>
+			<exclusions>
+				<exclusion>
+					<groupId>asm</groupId>
```
Comment (StephanEwen) [inline flink-shaded-hadoop/flink-shaded-include-yarn/pom.xml]: I this groupId correct, or should it be `org.ow2.asm` ?

File: flink-shaded-hadoop/flink-shaded-include-yarn/pom.xml [hunk 7]
```diff
@@ -0,0 +1,286 @@
...
+			</exclusions>
+		</dependency>
+
+		<dependency>
+			<groupId>org.apache.hadoop</groupId>
+			<artifactId>hadoop-mapreduce-client-core</artifactId>
+			<version>${hadoop.version}</version>
+			<exclusions>
+				<exclusion>
+					<groupId>asm</groupId>
```
Comment (StephanEwen) [inline flink-shaded-hadoop/flink-shaded-include-yarn/pom.xml]: I this groupId correct, or should it be `org.ow2.asm` ?

File: flink-shaded-hadoop/flink-shaded-include-yarn/pom.xml [hunk 8]
```diff
@@ -0,0 +1,286 @@
...
+			</exclusions>
+		</dependency>
+
+		<dependency>
+			<groupId>org.apache.hadoop</groupId>
+			<artifactId>hadoop-yarn-client</artifactId>
+			<version>${hadoop.version}</version>
+			<exclusions>
+				<exclusion>
+					<groupId>asm</groupId>
```
Comment (StephanEwen) [inline flink-shaded-hadoop/flink-shaded-include-yarn/pom.xml]: I this groupId correct, or should it be `org.ow2.asm` ?

File: flink-shaded-hadoop/flink-shaded-include-yarn/pom.xml [hunk 9]
```diff
@@ -0,0 +1,286 @@
...
+			</exclusions>
+		</dependency>
+
+		<dependency>
+			<groupId>org.apache.hadoop</groupId>
+			<artifactId>hadoop-yarn-common</artifactId>
+			<version>${hadoop.version}</version>
+			<exclusions>
+				<exclusion>
+					<groupId>asm</groupId>
```
Comment (StephanEwen) [inline flink-shaded-hadoop/flink-shaded-include-yarn/pom.xml]: I this groupId correct, or should it be `org.ow2.asm` ?

**Comment by StephanEwen** [comment]
Here are some points to double check
  - You exclude all the `flink-clients` web docs from the uberjar. I would guess this breaks the webclient, because it uses the JAR file as its web root, or am I overlooking something?
  - The uberjar has the YARN session as the main class. Seems strange to make this one the main class of the Uberjar, when there are several main classes (JobManager, TaskManager, CliFrontend, WebClient, AppMaster, ...)
  - Previously, the shade plugin was deactivated for the quickstarts. Is that no longer necessary?
  - Comments inline about the `asm` groupId

If none of these points is actually an issue, then +1 to merge

**Comment by rmetzger** [comment]
Thank you for reviewing the change.

* The exclusions for the `flink-clients` were indeed wrong. Good catch!
* It doesn't make much sense to set a main class for the uberjar. We don't use that property anywhere. I removed it.
* I've deactivated shading again for the quickstarts ..
* It seems that the `asm` package changed the groupId around the `4.0` release. I'll add both groupIds to the exclusions.

I've updated the pull requests. Lets see what travis says about it: https://travis-ci.org/rmetzger/flink/builds/53795738

**Comment by StephanEwen** [comment]
Looks good to me.

+1 to merge, if Travis (or Bob the Builder) agrees

File: flink-shaded-hadoop/flink-shaded-include-yarn-tests/pom.xml [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+
+	<packaging>jar</packaging>
+
+	<dependencies>
+		<dependency>
+			<groupId>org.apache.hadoop</groupId>
+			<artifactId>hadoop-common</artifactId>
+			<version>${hadoop.version}</version>
+		</dependency>
+		<!-- NOTE: We do not exclude hadoop's dependencies so that the tests start property (in particular jetty) -->
```
Comment (tillrohrmann) [inline flink-shaded-hadoop/flink-shaded-include-yarn-tests/pom.xml]: typo: properly

File: flink-shaded-hadoop/pom.xml [hunk 11]
```diff
@@ -60,19 +73,36 @@ under the License.
...
+								<transformer implementation="org.apache.maven.plugins.shade.resource.ServicesResourceTransformer"/>
+							</transformers>
 							<relocations>
 								<relocation>
 									<pattern>com.google</pattern>
-									<shadedPattern>org.apache.flink.shaded.com.google</shadedPattern>
+									<shadedPattern>org.apache.flink.hadoop.shaded.com.google</shadedPattern>
 									<excludes>
-										<exclude>com.google.protobuf.**</exclude>
+										<!-- <exclude>com.google.protobuf.**</exclude> -->
```
Comment (tillrohrmann) [inline flink-shaded-hadoop/pom.xml]: Why commenting out and not removing?

File: flink-yarn/src/main/java/org/apache/flink/yarn/FlinkYarnClient.java [hunk 12]
```diff
@@ -238,6 +239,15 @@ public int getTaskManagerCount() {
 
 	@Override
 	public void setShipFiles(List<File> shipFiles) {
+		File shipFile;
+		for(Iterator<File> it = shipFiles.iterator(); it.hasNext(); ) {
+			shipFile = it.next();
+			// remove uberjar from ship list (by default everything in the lib/ folder is added to
+			// the list of files to ship, but we handle the uberjar separatly.
```
Comment (tillrohrmann) [inline flink-yarn/src/main/java/org/apache/flink/yarn/FlinkYarnClient.java]: typo: separately

File: pom.xml [hunk 13]
```diff
@@ -1159,6 +726,55 @@ under the License.
...
+						<configuration>
+							<shadeTestJar>true</shadeTestJar>
+							<shadedArtifactAttached>false</shadedArtifactAttached>
+							<createDependencyReducedPom>true</createDependencyReducedPom>
+							<dependencyReducedPomLocation>${project.basedir}/target/dependency-reduced-pom.xml</dependencyReducedPomLocation>
+							<artifactSet>
+								<includes>
+									<include>com.google.guava:*</include>
+									<include>org.ow2.asm:*</include>
+									<!--  <include>org.apache.flume:*</include> Required because flume depends on guava. This only affects the flink-streaming connector which is not shipped through flink-dist by default. So its only about users adding flume as a dependency. -->
```
Comment (tillrohrmann) [inline pom.xml]: Why is it commented out rather than removed?

**Comment by tillrohrmann** [comment]
That are some massive changes to the pom files. As far as I can tell it looks good to me. The ```Akka``` changes look good.

**Comment by rmetzger** [comment]
Thanks for the review.
I'll address your comments and then merge it.

**Comment by hsaputra** [comment]
How much time does this new shading add to the total compile? It used to be around 16-18mins for me using mvn clean install -DskipTests.

I just did merge today and it has been more than 25 mins and has not complete the build =(

**Comment by rmetzger** [comment]
Hey Henry,
I'm sorry that the change is making your builds so much slower. My linux machine needs 5 minutes to build the current master.
I suspect there is something wrong on your MBP setup.

**Pull request closed without merging**